common/metrics/fx.go

```diff
@@ -0,0 +1,56 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package metrics
+
+import (
+	"context"
+
+	"go.uber.org/fx"
+)
+
+var RuntimeMetricsReporterModule = fx.Options(
+	RuntimeMetricsReporterLifetimeHooksModule,
+)
+var RuntimeMetricsReporterLifetimeHooksModule = fx.Options(
+	fx.Invoke(RuntimeMetricsReporterLifetimeHooks),
+)
+
+func RuntimeMetricsReporterLifetimeHooks(
+	lc fx.Lifecycle,
+	reporter *RuntimeMetricsReporter,
+) {
```

Comment:
Would it make sense to have the reporter provided at the same module where the start/stop hook is registered?

Reply:
Monitor is an interface that can start and close, but implementations are defined in different packages. As a further improvement, we can look into changing the instantiation of relevant objects to include this hook in fx module. I was thinking of moving these hooks to Daemon, but that would be even more indirect.

Reply:
Similar to above: we always want to start/stop interface, but init it in different locations. I would prefer not to copy-paste this code.